android/src/main/java/com/pagopa/securestorage/SecureStorage.kt

```diff
@@ -0,0 +1,400 @@
+import android.os.Build
+import android.os.storage.StorageManager
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyInfo
+import android.security.keystore.KeyProperties
+import android.security.keystore.KeyProperties.SECURITY_LEVEL_STRONGBOX
+import android.security.keystore.KeyProperties.SECURITY_LEVEL_TRUSTED_ENVIRONMENT
+import android.security.keystore.KeyProperties.SECURITY_LEVEL_UNKNOWN_SECURE
+import android.util.AtomicFile
+import java.io.ByteArrayOutputStream
+import java.io.File
+import java.io.FileNotFoundException
+import java.io.FileOutputStream
+import java.io.UnsupportedEncodingException
+import java.net.URLDecoder
+import java.net.URLEncoder
+import java.security.KeyStore
+import javax.crypto.Cipher
+import javax.crypto.KeyGenerator
+import javax.crypto.SecretKey
+import javax.crypto.SecretKeyFactory
+import javax.crypto.spec.GCMParameterSpec
+
+
+/**
+ * A secure storage class for writing on encrypted files.
+ * This class has a private constructor, thus instances must be created by using its [Builder].
+ * @property context the application context.
+ * @property storageDirectory the directory where the files will be stored.
+ * @property useEncryption if manual encryption has to be used.
+ */
+class SecureStorage private constructor(
+  context: Context, storageDirectory: File, useEncryption: Boolean
+) {
+
+  private val context: Context
+  private val storageDirectory: File
+  private val useEncryption: Boolean
+
+  init {
+    this.context = context
+    this.storageDirectory = storageDirectory
+    this.useEncryption = useEncryption
+  }
+
+  /**
+   * Removes the file associated with the given [key].
+   * @param key the identifier of the file.
+   * @throws SecureStorageException if an error occurs while deleting the file.
+   */
+  fun remove(key: String) {
+    try {
+      val atomicFile = AtomicFile(getFile(key))
+      atomicFile.delete()
+    } catch (e: Exception) {
+      throw SecureStorageException("An error occurred while deleting the file", e)
+    }
+  }
+
+  /**
+   * Clears [storageDirectory] from every file.
+   */
+  fun clear() {
+    storageDirectory.listFiles()?.let { fileList ->
+      fileList
+        .filter { it.name.startsWith(PREFIX) }
+        .forEach { it.delete() }
+    }
+  }
+
+  /**
+   * Enumerates keys for each file in [storageDirectory].
+   * @return a list containing each file name UTF-8 encoded.
+   * @throws SecureStorageException when UTF-8 encoding is not supported.
+   */
+  fun keys(): List<String> {
+    return storageDirectory.listFiles()?.filter { it.name.startsWith(PREFIX) }
+      ?.map {
+        try {
+          URLDecoder.decode(it.name.substring(PREFIX.length), "UTF-8")
+        } catch (e: UnsupportedEncodingException) {
+          throw SecureStorageException(
+            "UTF-8 encoding is not supported",
+            e
+          )
+        }
+      }
+      ?.toList() ?: emptyList()
+  }
```

Comment:
```suggestion
  fun keys(): List<String> {
    return storageDirectory.listFiles()?.filter { it.name.startsWith(PREFIX) }
      ?.map {file ->
        try {
          URLDecoder.decode(file.name.substring(PREFIX.length), "UTF-8")
        } catch (e: UnsupportedEncodingException) {
          throw SecureStorageException(
            "UTF-8 encoding is not supported",
            e
          )
        }
      } ?: emptyList()
  }
```

Reply:
I guess the cast isn't necessary at all, thank you, nice catch.